File: app/src/main/java/com/example/recyclerview/ViewHolder.java
package com.example.recyclerview;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.View;

public abstract class ViewHolder<Model> extends RecyclerView.ViewHolder
{
    private final View view;

    private Model model;

    protected ViewHolder(View itemView)
    {
        super(itemView);
        view = itemView;
    }

    public final View getView()
    {
        return view;
    }

    public final Model getModel()
    {
        return model;
    }

    public final void setModel(Model model)
    {
        this.model = model;
        onSetModel(model);
    }

    protected abstract void onSetModel(Model newModel);

    protected final Context getContext()
    {
        return view.getContext();
    }
}

File: app/src/main/java/com/example/recyclerview/ViewHolder.kt
package com.example.recyclerview

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.View

abstract class ViewHolder<Model> protected constructor(val view:View):RecyclerView.ViewHolder(view)
{
    var model:Model? = null

        set(value)
        {
            field = value
            onSetModel(value!!)
        }

    protected abstract fun onSetModel(model:Model)

    protected val context:Context get() = view.context
}
